halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/services/v1/DeploymentEnvironmentService.java

```diff
@@ -56,14 +62,26 @@ public DeploymentEnvironment getDeploymentEnvironment(String deploymentName) {
 
   public void setDeploymentEnvironment(String deploymentName, DeploymentEnvironment newDeploymentEnvironment) {
     DeploymentConfiguration deploymentConfiguration = deploymentService.getDeploymentConfiguration(deploymentName);
+
+    String defaultLocation = lookupDefaultLocation(deploymentName, newDeploymentEnvironment.getAccountName());
+    String location = newDeploymentEnvironment.getLocation();
+    newDeploymentEnvironment.setLocation(StringUtils.isEmpty(location) ? defaultLocation : location);
```

Comment:
Sorry, I wasn't clear - I meant each consecutive change that switches the provider with the account change.

In your example, the correct behavior would be to use the default location for openstack during deploy time.

I see why you want to keep the namespace explicit - but if a user wants to make the location explicit they are welcome to do so in the config, and once they do, the location sticks.

There are two reasons why I don't like having the location set on the first edit to DeploymentEnvironment:

1. The way it's currently configured lends itself to weird side-effects implicitly tied to the order your carry out certain operations. If you first decide to deploy to GCE, then K8s, you're deploying to a different namespace than you would if you had just selected K8s in the first place.

2. While we should have proper validation of the deployment environment that checks that the specified location is valid, not all providers (e.g. Kubernetes) will be able to catch accidents where a user doesn't update the location. Techincally "us-central1-f" is a valid k8s namespace, but it's kind-of sloppy.